addressing: describe multiaddrs containging the security protocol (#353)
libp2p/specs: addressing/README.md

unsurprising if you consider the utility of the `/ip4/7.7.7.7/ws` address that
would result from simply removing the `tcp` component.

> ### The security multiaddr

Reviewer: ```suggestion
### The multiaddr security component
```

While I know that each multiaddr component (e.g. `/ip6/::1` or `/tcp/1234`) in itself is a valid multiaddr, I don't think there is much value in a single component `/tls` multiaddr. Thus instead of speaking of _the security multiaddr_, I would refer to something along the lines of the _security multiaddr component_.

Author: Good point.

> Peers MAY advertise their addresses without the security protocol, e.g.

Reviewer: ```suggestion
Peers MAY advertise their addresses without a security protocol, e.g.
```

There are multiple security protocols, e.g. `tls` and `noise`, thus using `the` instead of `a` seems off here. I am not a native speaker, thus my intuition might be wrong.

`/ip4/6.6.6.6/tcp/1234/` or `/ip4/6.6.6.6/udp/1234/quic`. The security handshake

> protocol is then negotiated using the [multistream](../multistream.md). This is

Reviewer: ```suggestion
protocol is then negotiated using [multistream-select](../connections/README.md#multistream-select). This is
```

the way the libp2p handshake worked until mid 2021.
This poses a security problem, as the negotiation was not authenticated and
therefore susceptible to man-in-the-middle attacks. A MITM could modify the list
of supported handshake protocols, thereby forcing a downgrade to a (potentially)
less secure handshake protocol. Note that since QUIC is standardized to use
TLS 1.3, no handshake protocol needs to be negotiated when using QUIC.

> Peers SHOULD encapsulate the security protocol in the addresses they advertise,

Reviewer: Nit pick: Given that this is the section about security protocols in multiaddr, wouldn't it make more sense to have this paragraph and the next go first, especially as multiaddr without security protocols will not be used for much longer once this is in place?

Author: Where exactly? Do you mean before Encapsulation?

Reviewer: Does the following diff improve my suggestion?

```diff
diff --git a/addressing/README.md b/addressing/README.md
index abc6bc5..0967407 100644
--- a/addressing/README.md
+++ b/addressing/README.md
@@ -149,16 +149,6 @@ would result from simply removing the `tcp` component.
 
 ### The multiaddr security component
 
-Peers MAY advertise their addresses without a security protocol, e.g.
-`/ip4/6.6.6.6/tcp/1234/` or `/ip4/6.6.6.6/udp/1234/quic`. The security handshake
-protocol is then negotiated using [multistream-select](../connections/README.md#multistream-select). This is
-the way the libp2p handshake worked until mid 2021.
-This poses a security problem, as the negotiation was not authenticated and
-therefore susceptible to man-in-the-middle attacks. A MITM could modify the list
-of supported handshake protocols, thereby forcing a downgrade to a (potentially)
-less secure handshake protocol. Note that since QUIC is standardized to use
-TLS 1.3, no handshake protocol needs to be negotiated when using QUIC.
-
 Peers SHOULD encapsulate the security protocol in the addresses they advertise,
 for example `/ip4/6.6.6.6/tcp/1234/tls` for a TLS 1.3 server listening on TCP
 port 1234 and `/ip4/6.6.6.6/tcp/1235/noise` for a Noise server listening on TCP
@@ -180,6 +170,15 @@ Implementations using [Protocol Select](https://github.com/libp2p/specs/pull/349
 Note that it’s not valid to assume that any node that encapsulated the security
 protocol in their multiaddr also supports Protocol Select.
 
+Peers MAY advertise their addresses without a security protocol, e.g.
+`/ip4/6.6.6.6/tcp/1234/` or `/ip4/6.6.6.6/udp/1234/quic`. The security handshake
+protocol is then negotiated using [multistream-select](../connections/README.md#multistream-select). This is
+the way the libp2p handshake worked until mid 2021.
+This poses a security problem, as the negotiation was not authenticated and
+therefore susceptible to man-in-the-middle attacks. A MITM could modify the list
+of supported handshake protocols, thereby forcing a downgrade to a (potentially)
+less secure handshake protocol. Note that since QUIC is standardized to use
+TLS 1.3, no handshake protocol needs to be negotiated when using QUIC.
 ```

for example `/ip4/6.6.6.6/tcp/1234/tls` for a TLS 1.3 server listening on TCP
port 1234 and `/ip4/6.6.6.6/tcp/1235/noise` for a Noise server listening on TCP
port 1235. QUIC multiaddrs remain unchanged.
The nodes jump straight into a cryptographic handshake, thus curtailing the
possibility of packet-inspection-based censorship and dynamic downgrade attacks.
This also applies to circuit addresses: the security protocol is encoded in the
`<destination address>` as defined in [`p2p-circuit` Relay Addresses](#p2p-circuit-relay-addresses).

> 

Reviewer: ```suggestion

Advertising the secure channel protocol through the peer's Multiaddr instead of
negotiating the protocol in-band forces users to advertise an updated Multiaddr
when changing the secure channel protocol in use. This is especially cumbersome
when using hardcoded Multiaddresses. Users may leverage the [dnsaddr] Multiaddr
protocol as well as using a new UDP or TCP port for the new protocol to ease the
transition.

```

What do you think of adopting this section from https://github.com/libp2p/specs/pull/354/?

Author: That's a good idea!

Implementations using [Protocol Select](https://github.com/libp2p/specs/pull/349/)
(**TODO**: update link) MUST encapsulate the security protocol in the multiaddr.
Note that it’s valid to assume that any node that encapsulated the security

> protocol in their multiaddr also supports Protocol Select.

Reviewer: I thought last consensus was on not tying the two together. Again, I am fine either way, I only want to be consistent.

https://github.com/libp2p/specs/pull/349 currently includes:

> Note: A peer MAY advertise a Multiaddr that includes a secure channel handshake
protocol like `/noise` even if it doesn't support Protocol Select. See
[Heuristic section](#heuristic) below for details on how listeners can
differentiate the negotiation protocol spoken by the dialer on incoming
connections.

Author: The intention is to not tie them together, in the sense that advertising a security-enabled multiaddr doesn't mean that you speak Protocol Select. You can advertise a security-enabled multiaddr and still speak multistream.

However, I thought we agreed to require you to advertise a security-enabled multiaddr if you want to speak Protocol Select. Maybe that's not necessary though and you could use the Protocol Select protobufs to negotiate your handshake protocol?

Reviewer: ### Protocol Select requiring secure multiaddr component

> However, I thought we agreed to require you to advertise a security-enabled multiaddr if you want to speak Protocol Select.

Yes. We agreed on that.

>  Maybe that's not necessary though and you could use the Protocol Select protobufs to negotiate your handshake protocol?

I don't see a benefit of supporting this. Anything I am missing?

### Secure multiaddr component implying Protocol Select

> in the sense that advertising a security-enabled multiaddr doesn't mean that you speak Protocol Select.

> Note that it’s valid to assume that any node that encapsulated the security protocol in their multiaddr also supports Protocol Select.

Are these two sentences not in direct conflict?

Reviewer: ### Protocol Select requiring secure multiaddr component

> However, I thought we agreed to require you to advertise a security-enabled multiaddr if you want to speak Protocol Select.

Yes. We agreed on that.

>  Maybe that's not necessary though and you could use the Protocol Select protobufs to negotiate your handshake protocol?

I don't see a benefit of supporting this. Anything I am missing?

### Secure multiaddr component implying Protocol Select

> in the sense that advertising a security-enabled multiaddr doesn't mean that you speak Protocol Select.

> Note that it’s valid to assume that any node that encapsulated the security protocol in their multiaddr also supports Protocol Select.

Are these two sentences not in direct conflict?

Author: There was a "not" missing... Fixed.

unsurprising if you consider the utility of the `/ip4/7.7.7.7/ws` address that
would result from simply removing the `tcp` component.

### The multiaddr security component

Peers MAY advertise their addresses without a security protocol, e.g.
`/ip4/6.6.6.6/tcp/1234/` or `/ip4/6.6.6.6/udp/1234/quic`. The security handshake
protocol is then negotiated using [multistream-select](../connections/README.md#multistream-select). This is
the way the libp2p handshake worked until mid 2021.
This poses a security problem, as the negotiation was not authenticated and
therefore susceptible to man-in-the-middle attacks. A MITM could modify the list
of supported handshake protocols, thereby forcing a downgrade to a (potentially)
less secure handshake protocol. Note that since QUIC is standardized to use
TLS 1.3, no handshake protocol needs to be negotiated when using QUIC.

Peers SHOULD encapsulate the security protocol in the addresses they advertise,
for example `/ip4/6.6.6.6/tcp/1234/tls` for a TLS 1.3 server listening on TCP
port 1234 and `/ip4/6.6.6.6/tcp/1235/noise` for a Noise server listening on TCP

> port 1235. QUIC multiaddrs remain unchanged.

Reviewer: Resolving here with @marten-seemann recent comment in mind. Feel free to open a new thread in case you want to continue the discussion @kpp.

   `p2p-circuit`).

   Note: Specifying this security protocol is only necessary for active
   relaying. In the case of passive relaying the connection established by the
   destination to the relay will be used to relay the connection.

   Example:
   - Passive relaying: `<relay-multiaddr>/p2p-circuit/p2p/QmDestination`
   - Active relaying: `<relay-multiaddr>/p2p-circuit/ip4/6.6.6.6/tcp/1234/tls/p2p/QmDestination`

3. Upgrading the relayed connection from the source to the destination.

   The security protocol is specified by appending
   `/p2p-circuit-inner/<relayed-connection-security-protocol>` to the full
   address.

   <!-- TODO: Is `p2p-circuit-inner` the ideal name? Up for alternative
   suggestions. -->

   Example: `<relay-mulitaddr>/p2p-circuit/<destination-multiaddr>/p2p-circuit-inner/tls`

   Note: One might be tempted to not specify (3) and simply use the security
   protocol in (2). This would break if the security protocol used for (2) can
   not be used for (3), e.g. in the case where the relay establishes a QUIC
   connection to the destination secured via TLS and the source only supports
   Noise.

> 

Reviewer: Good idea. Added via https://github.com/libp2p/specs/pull/353/commits/6d8004e9cd5ec58f8b564443554838a553c02f0e. Thanks @yusefnapora.

connection involves 3 security protocol upgrades:

1. Upgrading the connection from the source to the relay.

   The security protocol is specified in the relay multiaddr (before
   `p2p-circuit`).

   Example: `/ip4/6.6.6.6/tcp/1234/tls/p2p/QmRelay/p2p-circuit/<destination-multiaddr>`

2. Upgrading the connection from the relay to the destination.

   The security protocol is specified in the destination multiaddr (after
   `p2p-circuit`).

   Note: Specifying this security protocol is only necessary for active
   relaying. In the case of passive relaying the connection established by the
   destination to the relay will be used to relay the connection.

   Example:
   - Passive relaying: `<relay-multiaddr>/p2p-circuit/p2p/QmDestination`
   - Active relaying: `<relay-multiaddr>/p2p-circuit/ip4/6.6.6.6/tcp/1234/tls/p2p/QmDestination`

3. Upgrading the relayed connection from the source to the destination.

   The security protocol is specified by appending
   `/p2p-circuit-inner/<relayed-connection-security-protocol>` to the full
   address.

   <!-- TODO: Is `p2p-circuit-inner` the ideal name? Up for alternative
   suggestions. -->

>    Example: `<relay-mulitaddr>/p2p-circuit/<destination-multiaddr>/p2p-circuit-inner/tls`

Reviewer 2: This will require some changes in our our relay implementations before it can be effected, by I am fine with that.

I agree that using a new inner component is the way to go, but can we call it something else?
Perhaps `p2p-circuit-sec` or `p2p-circuit-termination`.

Reviewer: Renamed in https://github.com/libp2p/specs/pull/353/commits/3c1487a106943ade936c5185def680c67c9496c8.
